### PR TITLE
Btc bridge

### PR DESCRIPTION
**Pull request by baidang201**
https://github.com/ggxchain/planning/issues/206

### CODE REVIEW

File: runtime/brooklyn/src/btcbridge.rs [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+
+parameter_types! {
+  pub const GetNativeCurrencyId: CurrencyId = runtime_common::constants::currency::NATIVE_CURRENCY_ID;
+  pub const GetRelayChainCurrencyId: CurrencyId = runtime_common::constants::currency::PARENT_CURRENCY_ID;
+  pub const GetWrappedCurrencyId: CurrencyId = runtime_common::constants::currency::WRAPPED_CURRENCY_ID;
+}
+
+// Pallet accounts
+parameter_types! {
+  pub const FeePalletId: PalletId = PalletId(*b"mod/fees");
```
Comment (akorchyn) [inline runtime/brooklyn/src/btcbridge.rs:187]: is there a sense to add btc prefix ?

File: runtime/brooklyn/src/lib.rs [hunk 2]
```diff
@@ -2,7 +2,7 @@
 
 #![cfg_attr(not(feature = "std"), no_std)]
 // `construct_runtime!` does a lot of recursion and requires us to increase the limit to 256.
-#![recursion_limit = "512"]
+#![recursion_limit = "2048"]
```
Comment (akorchyn) [inline runtime/brooklyn/src/lib.rs:5]: oof, it's going up and up.

File: runtime/runtime-common/Cargo.toml [hunk 3]
```diff
@@ -46,6 +46,25 @@ pallet-receipt-registry = { workspace = true, optional = true }
 cfg-if.workspace = true
 log.workspace = true
 
+# Orml
+orml-asset-registry.workspace = true
```
Comment (akorchyn) [inline runtime/runtime-common/Cargo.toml:50]: I would prefer to feature gate it until we release it to sydney to not make sydney build heavier.

**Review by akorchyn — APPROVED** [review]
Small nitpicks that I would vote to fix. But in general, it's awesome work

File: runtime/brooklyn/src/btcbridge.rs [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+};
+
+pub fn get_all_module_accounts() -> Vec<AccountId> {
+	vec![] // todo for product env
+}
+
+pub struct DustRemovalWhitelist;
+impl Contains<AccountId> for DustRemovalWhitelist {
+	fn contains(_a: &AccountId) -> bool {
+		//get_all_module_accounts().contains(a) // todo for product env
```
[the remark this answers is not in the document]
Comment (baidang201) [reply runtime/brooklyn/src/btcbridge.rs:23]: we need dust removal, i will add it

File: runtime/brooklyn/src/btcbridge.rs (hunk 1 above)
Comment (baidang201) [reply runtime/brooklyn/src/btcbridge.rs:187]: do you mean the filename start with "btc"?

File: runtime/brooklyn/src/lib.rs (hunk 2 above)
Comment (baidang201) [reply runtime/brooklyn/src/lib.rs:5]: we can try 1024

File: runtime/brooklyn/Cargo.toml [hunk 5]
```diff
@@ -219,6 +252,35 @@ std = [
...
+	"issue/std",
+	"nomination/std",
+	"oracle/std",
+	"primitives/std",
+	"redeem/std",
+	"replace/std",
+	"reward/std",
+	"security/std",
+	"staking/std",
+	"vault-registry/std",
```
[the remark this answers is not in the document]
Comment (baidang201) [reply runtime/brooklyn/Cargo.toml:275]: ok

File: node/src/runtime/testnet.rs [hunk 6]
```diff
@@ -50,16 +55,46 @@ impl ValidatorIdentity {
...
 	nominate: bool,
+	bitcoin_confirmations: u32,
+	disable_difficulty_check: bool,
 ) -> GenesisConfig {
 	let mut rng = rand::rngs::StdRng::seed_from_u64(0);
 	let stash = 1000 * GGX;
+	const DEFAULT_MAX_DELAY_MS: u32 = 60 * 60 * 1000; // one hour
+	const DEFAULT_DUST_VALUE: Balance = 1000;
+
+	let block_time_in_millis = 2000;
```
[the remark this answers is not in the document]
Comment (baidang201) [reply node/src/runtime/testnet.rs:94]: i try find  error "Thread 'main' panicked at '`get_version_1` called outside of an Externalities-provided environment."
```
Version: 0.1.2-8822ae82eb1

   0: backtrace::backtrace::libunwind::trace
             at /Users/li/.cargo/registry/src/mirrors.ustc.edu.cn-61ef6e0cd06fb9b8/backtrace-0.3.69/src/backtrace/libunwind.rs:93:5
      backtrace::backtrace::trace_unsynchronized
             at /Users/li/.cargo/registry/src/mirrors.ustc.edu.cn-61ef6e0cd06fb9b8/backtrace-0.3.69/src/backtrace/mod.rs:66:5
   1: backtrace::backtrace::trace
             at /Users/li/.cargo/registry/src/mirrors.ustc.edu.cn-61ef6e0cd06fb9b8/backtrace-0.3.69/src/backtrace/mod.rs:53:14
   2: backtrace::capture::Backtrace::create
             at /Users/li/.cargo/registry/src/mirrors.ustc.edu.cn-61ef6e0cd06fb9b8/backtrace-0.3.69/src/capture.rs:176:9
   3: backtrace::capture::Backtrace::new
             at /Users/li/.cargo/registry/src/mirrors.ustc.edu.cn-61ef6e0cd06fb9b8/backtrace-0.3.69/src/capture.rs:140:22
   4: sp_panic_handler::panic_hook
             at /Users/li/.cargo/git/checkouts/substrate-7e08433d4c370a21/5e49f6e/primitives/panic-handler/src/lib.rs:166:18
   5: sp_panic_handler::set::{{closure}}
             at /Users/li/.cargo/git/checkouts/substrate-7e08433d4c370a21/5e49f6e/primitives/panic-handler/src/lib.rs:62:12
   6: <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call
             at /rustc/d06ca0ffaf4ac72732665f99dd9ad962194cd0b3/library/alloc/src/boxed.rs:2021:9
      std::panicking::rust_panic_with_hook
             at /rustc/d06ca0ffaf4ac72732665f99dd9ad962194cd0b3/library/std/src/panicking.rs:733:13
   7: std::panicking::begin_panic_handler::{{closure}}
             at /rustc/d06ca0ffaf4ac72732665f99dd9ad962194cd0b3/library/std/src/panicking.rs:621:13
   8: std::sys_common::backtrace::__rust_end_short_backtrace
             at /rustc/d06ca0ffaf4ac72732665f99dd9ad962194cd0b3/library/std/src/sys_common/backtrace.rs:170:18
   9: rust_begin_unwind
             at /rustc/d06ca0ffaf4ac72732665f99dd9ad962194cd0b3/library/std/src/panicking.rs:617:5
  10: core::panicking::panic_fmt
             at /rustc/d06ca0ffaf4ac72732665f99dd9ad962194cd0b3/library/core/src/panicking.rs:67:14
  11: core::panicking::panic_display
             at /rustc/d06ca0ffaf4ac72732665f99dd9ad962194cd0b3/library/core/src/panicking.rs:150:5
      core::panicking::panic_str
             at /rustc/d06ca0ffaf4ac72732665f99dd9ad962194cd0b3/library/core/src/panicking.rs:134:5
      core::option::expect_failed
             at /rustc/d06ca0ffaf4ac72732665f99dd9ad962194cd0b3/library/core/src/option.rs:1988:5
  12: core::option::Option<T>::expect
             at /rustc/d06ca0ffaf4ac72732665f99dd9ad962194cd0b3/library/core/src/option.rs:898:21
  13: sp_io::storage::get_version_1::{{closure}}
             at /Users/li/.cargo/git/checkouts/substrate-7e08433d4c370a21/5e49f6e/primitives/io/src/lib.rs:118:1
  14: tracing::span::Span::in_scope
             at /Users/li/.cargo/registry/src/mirrors.ustc.edu.cn-61ef6e0cd06fb9b8/tracing-0.1.37/src/span.rs:1102:9
  15: sp_io::storage::get_version_1
             at /Users/li/.cargo/git/checkouts/substrate-7e08433d4c370a21/5e49f6e/primitives/io/src/lib.rs:118:1
  16: sp_io::storage::get
             at /Users/li/.cargo/git/checkouts/substrate-7e08433d4c370a21/5e49f6e/primitives/io/src/lib.rs:118:1
  17: frame_support::storage::unhashed::get
             at /Users/li/.cargo/git/checkouts/substrate-7e08433d4c370a21/5e49f6e/frame/support/src/storage/unhashed.rs:25:2
  18: ggxchain_runtime_brooklyn::Hours::get
             at /Users/li/.cargo/git/checkouts/substrate-7e08433d4c370a21/5e49f6e/frame/support/src/lib.rs:384:5
  19: ggxchain_node::runtime::testnet::testnet_genesis
             at node/src/runtime/testnet.rs:279:15
  20: ggxchain_node::chain_spec::development_config::{{closure}}
             at node/src/chain_spec.rs:34:4
  21: sc_chain_spec::chain_spec::GenesisSource<G>::resolve
             at /Users/li/.cargo/git/checkouts/substrate-7e08433d4c370a21/5e49f6e/client/chain-spec/src/chain_spec.rs:82:44
  22: <sc_chain_spec::chain_spec::ChainSpec<G,E> as sp_runtime::BuildStorage>::assimilate_storage
             at /Users/li/.cargo/git/checkouts/substrate-7e08433d4c370a21/5e49f6e/client/chain-spec/src/chain_spec.rs:113:9
  23: sp_runtime::BuildStorage::build_storage
             at /Users/li/.cargo/git/checkouts/substrate-7e08433d4c370a21/5e49f6e/primitives/runtime/src/lib.rs:193:3
  24: sc_chain_spec::genesis::GenesisBlockBuilder<Block,B,E>::new
             at /Users/li/.cargo/git/checkouts/substrate-7e08433d4c370a21/5e49f6e/client/chain-spec/src/genesis.rs:113:4
  25: sc_service::builder::new_full_parts
             at /Users/li/.cargo/git/checkouts/substrate-7e08433d4c370a21/5e49f6e/client/service/src/builder.rs:146:30
  26: ggxchain_node::service::brooklyn::new_partial
             at node/src/service/brooklyn.rs:109:3
  27: ggxchain_node::service::brooklyn::new_full
             at node/src/service/brooklyn.rs:265:6
  28: ggxchain_node::command::run::{{closure}}::{{closure}}
             at node/src/command.rs:263:5
  29: tokio::runtime::park::CachedParkThread::block_on::{{closure}}
             at /Users/li/.cargo/registry/src/mirrors.ustc.edu.cn-61ef6e0cd06fb9b8/tokio-1.32.0/src/runtime/park.rs:282:63
  30: tokio::runtime::coop::with_budget
             at /Users/li/.cargo/registry/src/mirrors.ustc.edu.cn-61ef6e0cd06fb9b8/tokio-1.32.0/src/runtime/coop.rs:107:5
      tokio::runtime::coop::budget
             at /Users/li/.cargo/registry/src/mirrors.ustc.edu.cn-61ef6e0cd06fb9b8/tokio-1.32.0/src/runtime/coop.rs:73:5
      tokio::runtime::park::CachedParkThread::block_on
             at /Users/li/.cargo/registry/src/mirrors.ustc.edu.cn-61ef6e0cd06fb9b8/tokio-1.32.0/src/runtime/park.rs:282:31
  31: tokio::runtime::context::blocking::BlockingRegionGuard::block_on
             at /Users/li/.cargo/registry/src/mirrors.ustc.edu.cn-61ef6e0cd06fb9b8/tokio-1.32.0/src/runtime/context/blocking.rs:66:9
  32: tokio::runtime::scheduler::multi_thread::MultiThread::block_on::{{closure}}
             at /Users/li/.cargo/registry/src/mirrors.ustc.edu.cn-61ef6e0cd06fb9b8/tokio-1.32.0/src/runtime/scheduler/multi_thread/mod.rs:87:13
  33: tokio::runtime::context::runtime::enter_runtime
             at /Users/li/.cargo/registry/src/mirrors.ustc.edu.cn-61ef6e0cd06fb9b8/tokio-1.32.0/src/runtime/context/runtime.rs:65:16
  34: tokio::runtime::scheduler::multi_thread::MultiThread::block_on
             at /Users/li/.cargo/registry/src/mirrors.ustc.edu.cn-61ef6e0cd06fb9b8/tokio-1.32.0/src/runtime/scheduler/multi_thread/mod.rs:86:9
  35: tokio::runtime::runtime::Runtime::block_on
             at /Users/li/.cargo/registry/src/mirrors.ustc.edu.cn-61ef6e0cd06fb9b8/tokio-1.32.0/src/runtime/runtime.rs:349:45
  36: sc_cli::runner::Runner<C>::run_node_until_exit
             at /Users/li/.cargo/git/checkouts/substrate-7e08433d4c370a21/5e49f6e/client/cli/src/runner.rs:89:26
  37: ggxchain_node::command::run
             at node/src/command.rs:255:4
  38: ggxchain_node::main
             at node/src/main.rs:19:2
  39: core::ops::function::FnOnce::call_once
             at /rustc/d06ca0ffaf4ac72732665f99dd9ad962194cd0b3/library/core/src/ops/function.rs:250:5
  40: std::sys_common::backtrace::__rust_begin_short_backtrace
             at /rustc/d06ca0ffaf4ac72732665f99dd9ad962194cd0b3/library/std/src/sys_common/backtrace.rs:154:18
  41: std::rt::lang_start::{{closure}}
             at /rustc/d06ca0ffaf4ac72732665f99dd9ad962194cd0b3/library/std/src/rt.rs:166:18
  42: core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once
             at /rustc/d06ca0ffaf4ac72732665f99dd9ad962194cd0b3/library/core/src/ops/function.rs:284:13
      std::panicking::try::do_call
             at /rustc/d06ca0ffaf4ac72732665f99dd9ad962194cd0b3/library/std/src/panicking.rs:524:40
      std::panicking::try
             at /rustc/d06ca0ffaf4ac72732665f99dd9ad962194cd0b3/library/std/src/panicking.rs:488:19
      std::panic::catch_unwind
             at /rustc/d06ca0ffaf4ac72732665f99dd9ad962194cd0b3/library/std/src/panic.rs:142:14
      std::rt::lang_start_internal::{{closure}}
             at /rustc/d06ca0ffaf4ac72732665f99dd9ad962194cd0b3/library/std/src/rt.rs:148:48
      std::panicking::try::do_call
             at /rustc/d06ca0ffaf4ac72732665f99dd9ad962194cd0b3/library/std/src/panicking.rs:524:40
      std::panicking::try
             at /rustc/d06ca0ffaf4ac72732665f99dd9ad962194cd0b3/library/std/src/panicking.rs:488:19
      std::panic::catch_unwind
             at /rustc/d06ca0ffaf4ac72732665f99dd9ad962194cd0b3/library/std/src/panic.rs:142:14
      std::rt::lang_start_internal
             at /rustc/d06ca0ffaf4ac72732665f99dd9ad962194cd0b3/library/std/src/rt.rs:148:20
  43: std::rt::lang_start
             at /rustc/d06ca0ffaf4ac72732665f99dd9ad962194cd0b3/library/std/src/rt.rs:165:17
  44: _main


Thread 'main' panicked at '`get_version_1` called outside of an Externalities-provided environment.', /Users/li/.cargo/git/checkouts/substrate-7e08433d4c370a21/5e49f6e/primitives/io/src/lib.rs:120

This is a bug. Please report it at:

        support.anonymous.an
```

File: node/src/runtime/testnet.rs [hunk 7]
```diff
@@ -224,5 +259,105 @@ pub fn testnet_genesis(
...
 			phantom: std::marker::PhantomData,
 		},
+		asset_registry: Default::default(),
+		tokens: TokensConfig {
+			balances: endowed_accounts
+				.iter()
+				.flat_map(|k| {
+					vec![
+						(k.clone().0, Token(INTR), 1 << 60),
+						(k.clone().0, Token(DOT), 1 << 60),
```
[the remark this answers is not in the document]
Comment (baidang201) [reply node/src/runtime/testnet.rs:269]: we will use GGX later

File: runtime/brooklyn/Cargo.toml (hunk 5 above)
[the remark this answers is not in the document]
Comment (baidang201) [reply runtime/brooklyn/Cargo.toml:275]: fixed

File: runtime/brooklyn/src/btcbridge.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (baidang201) [reply runtime/brooklyn/src/btcbridge.rs]: fixed

File: runtime/brooklyn/src/btcbridge.rs [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+		true
+	}
+}
+
+parameter_types! {
+	pub const MaxLocks: u32 = 50;
```
[the remark this answers is not in the document]
Comment (baidang201) [reply runtime/brooklyn/src/btcbridge.rs]: fixed

File: runtime/brooklyn/src/lib.rs (hunk 2 above)
Comment (baidang201) [reply runtime/brooklyn/src/lib.rs]: fixed

File: runtime/runtime-common/src/constants.rs [hunk 9]
```diff
@@ -17,4 +17,10 @@ pub mod currency {
 	pub const fn deposit(items: u32, bytes: u32) -> Balance {
 		items as Balance * 2_000 * CENTS + (bytes as Balance) * 100 * MILLICENTS
 	}
+
+	/// Constant values used within the runtime.
+	pub const MILLIGGX: Balance = 1_000_000_000_000_000;
```
Comment (akorchyn) [inline runtime/runtime-common/src/constants.rs:22]: Why do we need two impls for UNITS and for MILLIGGX. Alogngside with two deposit functions?

Comment (baidang201) [reply runtime/runtime-common/src/constants.rs:22]: no, i will remove UNITS and merge deposit function

**Pull request merged**